homework/Long_Nguyen/lesson_7/HouseRobberIII.java

```diff
@@ -0,0 +1,31 @@
+import java.util.HashMap;
+import java.util.Map;
+
+// https://leetcode.com/problems/house-robber-iii
+public class HouseRobberIII {
+    public int rob(TreeNode root) {
+        return solver(root, new HashMap<>());
+    }
+
+    private int solver(TreeNode root, Map<TreeNode, Integer> dp) {
```

Comment:
will it be better if we declare dp as global variable instead of passing to `solver`?

my shorter solution:


```
class Solution {
public:
    vector<int> dfs(TreeNode* u) {
        if (!u) return {0, 0};
        auto left = dfs(u->left);
        auto right = dfs(u->right);
        int rob = u->val + left[1] + right[1];
        int notRob = max(left[0], left[1]) + max(right[0], right[1]);
        return {rob, notRob};
    }
    
    int rob(TreeNode* root) {
        auto res = dfs(root);
        return max(res[0], res[1]);
    }
};
```